src/rgw/rgw_sal_motr.cc

```diff
@@ -1933,8 +1933,19 @@ int MotrObject::MotrReadOp::prepare(optional_yield y, const DoutPrefixProvider*
   if (rc < 0)
     return rc;
 
-  if (ent.is_delete_marker())
-    return -ENOENT;
+  req_state* s = static_cast<req_state*>(rctx->get_private());
+
+  // In GET/HEAD object API, return "MethodNotAllowed"
+  // if delete-marker is the latest object entry
+  // Else, return "NoSuchKey" error
+  if (ent.is_delete_marker()) {
+    if (source->get_instance() == ent.key.instance && ent.key.instance != "") {
+      s->err.message = "The specified method is not allowed against this resource.";
+      return -ERR_METHOD_NOT_ALLOWED;
+    }
+    else
+      return -ENOENT;
```

Comment:
You don't need `else` here
```suggestion
    return -ENOENT;
```

Reply:
> You don't need `else` here

else is needed still, to keep ENOENT response for other regular cases. Here code diff is not reflecting properly as i see, here is the screenshot changed code
![image](https://user-images.githubusercontent.com/32053979/189288559-f27f2ee2-52cb-413f-9189-5a24f38d1c54.png)

Reply:
I meant, `return -ENOENT;` is required but no need to keep it inside `else`. 

These two code snippets will give the same results:
```
{
  if (something) {
    return this;
  }
  else
    return that;
}
```
```
{
  if (something) {
    return this;
  }
  return that;
}
```

Reply:
Done. Followed 2nd approach.